### PR TITLE
Fix function schemas

### LINKED ISSUE #882: UDFs may be in public schema on fresh install
Depending on whether you create extension or alter extension, some UDFs may be in a different schema.

```
postgres=# CREATE EXTENSION citus VERSION '5.2-1';
CREATE EXTENSION 
postgres=# ALTER EXTENSION citus UPDATE;
ALTER EXTENSION
postgres=# \df
                       List of functions
 Schema | Name | Result data type | Argument data types | Type 
--------+------+------------------+---------------------+------
(0 rows)
postgres=# CREATE EXTENSION citus;
CREATE EXTENSION
postgres=# \df
                                             List of functions
 Schema |             Name              | Result data type |          Argument data types          |  Type  
--------+-------------------------------+------------------+---------------------------------------+--------
 public | column_name_to_column         | text             | table_name regclass, column_name text | normal
 public | master_expire_table_cache     | void             | table_name regclass                   | normal
 public | master_get_new_placementid    | bigint           |                                       | normal
 public | worker_drop_distributed_table | void             | logicalrelid oid                      | normal
(4 rows)
```

### PR DESCRIPTION
**Pull request by jasonmp85**
We'd been relying on a single `SET search_path` command in an earlier script, but a subsequent script `RESET search_path`, causing any further bare functions to be created in the first schema on the search path.

However, starting with an older extension version and executing `ALTER` scripts one at a time **does** avoid putting any functions in the public namespace, so I wrote my own upgrade script by  specifying a `search_path` and not qualifying function names in my `ALTER FUNCTION` statements.

Fixes #882

### CODE REVIEW

File: src/test/regress/sql/multi_extension.sql [hunk 1]
```diff
@@ -44,6 +55,7 @@ ALTER EXTENSION citus UPDATE TO '6.0-10';
 ALTER EXTENSION citus UPDATE TO '6.0-11';
 ALTER EXTENSION citus UPDATE TO '6.0-12';
 ALTER EXTENSION citus UPDATE TO '6.0-13';
+ALTER EXTENSION citus UPDATE TO '6.0-14';
```
Comment (anarazel) [inline src/test/regress/sql/multi_extension.sql]: We should do the schema check after the upgrade scripts as well, that way we check whether they individually also do the correct thing, or whether they rely on earlier state.

Comment (jasonmp85) [inline src/test/regress/sql/multi_extension.sql]: So the interesting part is that the upgrade script I'd originally written failed because it explicitly used `public.<function_name>`. A bare `CREATE EXTENSION` apparently created those "bad" functions in the `public` schema (because of the `RESET search_path` call), but individual `ALTER EXTENSION` calls look as though they run inheriting the `search_path` of the caller maybe?!?

The immediate issue here is that I think we should devise some sort of pattern for writing our upgrade files (do we _always_ wrap them in a `SET search_path TO pg_catalog`/`RESET search_path`?)…

The lingering question of mine that this brings up is I've never really understood why we put anything in `pg_catalog` in the first place. Sure, there are historic reasons for it, but why not be a fully relocatable extension, or create a `citus` schema and put everything there? That's always bothered me, TBH.

**Comment by jasonmp85** [comment]
Addressed @anarazel's feedback.

File: src/test/regress/expected/multi_extension.out [hunk 2]
```diff
@@ -39,6 +55,23 @@ ALTER EXTENSION citus UPDATE TO '6.0-10';
...
+FROM pg_depend AS pgd,
+	 pg_extension as pge,
+	 pg_proc as pgp
+WHERE pgd.classid    = 'pg_proc'::regclass AND
+	  pgd.refclassid = 'pg_extension'::regclass AND
+	  pgd.objid      = pgp.oid AND
+	  pgd.refobjid   = pge.oid AND
+	  pge.extname    = 'citus' AND
+	  pgp.pronamespace != 'pg_catalog'::regnamespace;
+ count
```
Comment (anarazel) [inline src/test/regress/expected/multi_extension.out]: Hm. I don't quite see why we want to restrict this to pg_proc - shouldn't we warn about any references outside pg_catalog?

**Comment by aamederen** [comment]
Regarding to the unexpected `function recover_prepared_transactions() is already in schema "pg_catalog"` message, I have made some experiments:

If I create Citus extension directly from version 6.0-10 or beyond (where `recover_prepared_transactions()` UDF has been added), the function is created in public schema:
```
DROP EXTENSION citus;
CREATE EXTENSION citus VERSION '6.0-10';
\df recover_prepared_transactions
                                    List of functions
 Schema |             Name              | Result data type | Argument data types |  Type  
--------+-------------------------------+------------------+---------------------+--------
 public | recover_prepared_transactions | integer          |                     | normal
(1 row)

```

However, if I create Citus extension with version 5.0 and update it one-by-one, it ends up created in pg_catalog schema:
```
DROP EXTENSION citus;
CREATE EXTENSION citus VERSION '5.0';
ALTER EXTENSION citus UPDATE TO '5.0-1';
ALTER EXTENSION citus UPDATE TO '5.0-2';
ALTER EXTENSION citus UPDATE TO '5.1-1';
ALTER EXTENSION citus UPDATE TO '5.1-2';
ALTER EXTENSION citus UPDATE TO '5.1-3';
ALTER EXTENSION citus UPDATE TO '5.1-4';
ALTER EXTENSION citus UPDATE TO '5.1-5';
ALTER EXTENSION citus UPDATE TO '5.1-6';
ALTER EXTENSION citus UPDATE TO '5.1-7';
ALTER EXTENSION citus UPDATE TO '5.1-8';
ALTER EXTENSION citus UPDATE TO '5.2-1';
ALTER EXTENSION citus UPDATE TO '5.2-2';
ALTER EXTENSION citus UPDATE TO '5.2-3';
ALTER EXTENSION citus UPDATE TO '5.2-4';
ALTER EXTENSION citus UPDATE TO '6.0-1';
ALTER EXTENSION citus UPDATE TO '6.0-2';
ALTER EXTENSION citus UPDATE TO '6.0-3';
ALTER EXTENSION citus UPDATE TO '6.0-4';
ALTER EXTENSION citus UPDATE TO '6.0-5';
ALTER EXTENSION citus UPDATE TO '6.0-6';
ALTER EXTENSION citus UPDATE TO '6.0-7';
ALTER EXTENSION citus UPDATE TO '6.0-8';
ALTER EXTENSION citus UPDATE TO '6.0-9';
ALTER EXTENSION citus UPDATE TO '6.0-10';
\df recover_prepared_transactions
                                      List of functions
   Schema   |             Name              | Result data type | Argument data types |  Type  
------------+-------------------------------+------------------+---------------------+--------
 pg_catalog | recover_prepared_transactions | integer          |                     | normal
```

I could not understand the reason for this behavior.

**Comment by jasonmp85** [comment]
> Regarding to the unexpected […] message

Yup: see PR description, I saw the same thing.

**Comment by mtuncer** [comment]
installed citus v5.0.0
installed the fix on top of that
```
alter extension citus update;
ERROR:  function recover_prepared_transactions() is already in schema "pg_catalog"
\dx
                 List of installed extensions
  Name   | Version |   Schema   |         Description          
---------+---------+------------+------------------------------
 citus   | 5.0     | pg_catalog | Citus distributed database
 plpgsql | 1.0     | pg_catalog | PL/pgSQL procedural language
(2 rows)

```

**Comment by mtuncer** [comment]
It also happens in upgrade from 5.1 to current fix. Stopping further tests as this is already a showstopper. We still do not know the root cause of the UDFs different placement behavior.

We should perhaps just make sure UDF is placed into pg_catalog regardless of the original placement without erroring out at the time being.

**Comment by jasonmp85** [comment]
> We still do not know the root cause of the UDFs different placement behavior.

Uh, we do, actually?

**Comment by jasonmp85** [comment]
OK, sorry about the 9.5 failures. I've rewritten the script to explicitly move everything from `public` to `pg_catalog`, and added an `EXCEPTION` block that will short-circuit if the functions aren't found in `public` (i.e. if they're already in `pg_catalog`).

In theory, I could have an individual `EXCEPTION` block for every `ALTER FUNCTION` statement, but in practice if the first is not in `public`, neither are any subsequent ones. Besides, if we miss a function, the new test would fail anyways, letting us know.

@anarazel, I expanded the counting to apply to all objects, not just functions. I found that we do have a data type in the `citus` schema, so I whitelisted that one as well. Now we're counting all objects not in `pg_catalog` or `citus` and erroring if that count is not zero. And I figured out I needed a `LATERAL` in my query, which was neat.

File: src/backend/distributed/citus--6.0-13--6.0-14.sql [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+	SET SCHEMA pg_catalog;
+
+	ALTER FUNCTION public.master_get_new_placementid()
+	SET SCHEMA pg_catalog;
+
+	ALTER FUNCTION public.master_expire_table_cache(table_name regclass)
+	SET SCHEMA pg_catalog;
+
+	-- some installations don't need this corrective, so just skip...
+	EXCEPTION WHEN undefined_function THEN
```
Comment (anarazel) [inline src/backend/distributed/citus--6.0-13--6.0-14.sql]: Unimportant nitpick:  Imo this should be on the same level of the BEGIN, since that's actually its scope ;)

File: src/test/regress/expected/multi_extension.out [hunk 4]
```diff
@@ -7,6 +7,20 @@
 -- not done yet.
 ALTER SEQUENCE pg_catalog.pg_dist_shardid_seq RESTART 580000;
 ALTER SEQUENCE pg_catalog.pg_dist_jobid_seq RESTART 580000;
+-- ensure no functions were created outside pg_catalog
```
Comment (anarazel) [inline src/test/regress/expected/multi_extension.out]: s/functions/objects/

**Comment by jasonmp85** [comment]
@mtuncer, the reason executing the installation all at once puts things in `public` but executing things one-at-a-time does not is that there is a `RESET search_path` at the end of one of the files and each of the affected functions is defined without qualification (i.e. `CREATE FUNCTION foo` rather than `CREATE FUNCTION pg_catalog.foo`).

When extension install scripts run, they have their `search_path` set to the `schema` specified by the extension's control file. However, if a `RESET search_path` is used within the installation script, it reset to the session's default value, and _not_ to the `schema` from the control file.

Because of this, the `RESET` in the middle (with no subsequent `SET`) results in all unqualified further functions being created in `public`. But if you run the scripts one-at-a-time, each one starts with the `search_path` set to the extension's `schema` (in our case, `pg_catalog`), so they're safe.

In short, I'm starting to think we should consolidate our install files in the near future and excise all the `SET search_path`/`RESET search_path` stuff.

File: src/backend/distributed/citus--6.0-13--6.0-14.sql (hunk 3 above)
Comment (jasonmp85) [inline src/backend/distributed/citus--6.0-13--6.0-14.sql]: Ah, yeah I had it there but moved it over when I had a second `BEGIN` block for a brief moment. Fixing.

**Comment by jasonmp85** [comment]
Merging after this passes.

@anarazel, does it surprise you that `RESET search_path` inside of an extension script resets the path all the way back to whatever the originating session set, and not to simply the `schema` specified inside the `control` file (which is what `search_path` is initialized to upon first entering an extension SQL file)? Was unexpected to me.

**Comment by anarazel** [comment]
@jasonmp85 No, it does not surprise me, but perhaps that's because I know the implementation and everything. I wouldn't bet on that changing anytime soon.
